File: 2015/Day_17/Python/part_2.py
from sys import path
from itertools import combinations
path.insert(0, '../../input_parsing')
import parse_input

container_sizes = parse_input.input_per_line('../input.txt')
container_sizes = [int(x) for x in container_sizes]
all_ways = [element for size in range(1, len(container_sizes))
            for element in combinations(container_sizes, size) if sum(element) == 150]
minimum_number_of_containers = 999999999999
for combination in all_ways:
    if len(combination) < minimum_number_of_containers:
        minimum_number_of_containers = len(combination)
all_ways = [combination for combination in all_ways if len(combination) == minimum_number_of_containers]
print(f"The number of combinations for using my containers is {len(all_ways)}")

File: 2015/Day_17/Perl/part_2.pl
#!/usr/bin/perl

use v5.20;                                                                                     
use warnings;                                                                                  
use List::Util qw/sum/;
use  Data::Dump qw/dump/; 
use Algorithm::Combinatorics qw(combinations);


open(CONTAINER_SIZES, "../input.txt") || die "No containers!";

my @container_sizes = <CONTAINER_SIZES>;
# my @container_sizes = (20, 15, 10, 5, 5);
# my @container_sizes = ("20\n", "15\n", "10\n", "5\n", "5\n");


my $CONTAINER_SUM = 150;

my @all_ways;

my $size_all_ways = scalar @container_sizes;

foreach my $number (1.. $size_all_ways)
{
	my $permutations = combinations(\@container_sizes, $number);
	while (my $combination = $permutations->next)
	{
		if(sum(@$combination) == $CONTAINER_SUM)
		{
			push @all_ways, \@$combination;
		}	
	}
}

say "Done creating the containers from part 1....";

my $minimum_number_of_containers = 99999999999999999999;

foreach my $combination (@all_ways)
{
    if (scalar @$combination < $minimum_number_of_containers)
    {
        $minimum_number_of_containers = scalar @$combination;
    }
}

my @final_combinations;

foreach my $combination (@all_ways)
{
    if (scalar @$combination == $minimum_number_of_containers)
    {
        push @final_combinations,\@$combination;
    }
}

my $combo_numbers = scalar @final_combinations;
print "The number of minimum combinations for using my contains is $combo_numbers";
